src/gpio.c

```diff
@@ -0,0 +1,95 @@
+#include <util.h>
+#include <gpio.h>
+#include "main.h"
+
+void* gpio_perf(gpio_pin_t gpio_pin) {
+  if (gpio_pin & GPIO_A == GPIO_A) {
+    return GPIOA;
+  } else if (gpio_pin & GPIO_B == GPIO_B) {
+    return GPIOB;
+  } else if (gpio_pin & GPIO_C == GPIO_C) {
+    return GPIOC;
+  }
+}
+
+void gpio_clock(gpio_pin_t gpio_pin) {
+  /* RCC peripheral clock enable */
+  if (gpio_pin & GPIO_A == GPIO_A) {
+    RCC_AHBPeriphClockCmd(RCC_AHBPeriph_GPIOA, ENABLE);
+  } else if (gpio_pin & GPIO_B == GPIO_B) {
+    RCC_AHBPeriphClockCmd(RCC_AHBPeriph_GPIOB, ENABLE);
+  } else if (gpio_pin & GPIO_C == GPIO_C) {
+    RCC_AHBPeriphClockCmd(RCC_AHBPeriph_GPIOC, ENABLE);
+  }
+}
+
+void gpio_up(gpio_pin_t gpio_pin) {
+  GPIO_InitTypeDef GPIO_InitStructure;
+
+  gpio_clock(gpio_pin);
+
+  /* Configure pins in output pushpull mode */
+  GPIO_InitStructure.GPIO_Pin = (gpio_pin & GPIO_PIN_MASK);
+  GPIO_InitStructure.GPIO_Mode = GPIO_Mode_OUT;
+  GPIO_InitStructure.GPIO_OType = GPIO_OType_PP;
+  GPIO_InitStructure.GPIO_Speed = GPIO_Speed_50MHz;
+  GPIO_InitStructure.GPIO_PuPd = GPIO_PuPd_NOPULL;
+  GPIO_Init(GPIOC, &GPIO_InitStructure);
+}
+
+void gpio_on(gpio_pin_t gpio_pin) {
+  ((GPIO_TypeDef *)gpio_perf(gpio_pin))->BSRR = gpio_pin & GPIO_PIN_MASK;
+}
+
+void gpio_off(gpio_pin_t gpio_pin) {
+  ((GPIO_TypeDef *)gpio_perf(gpio_pin))->BRR = gpio_pin & GPIO_PIN_MASK;
+}
+
+void gpio_input(gpio_pin_t gpio_pin) {
+  EXTI_InitTypeDef   EXTI_InitStructure;
+  GPIO_InitTypeDef   GPIO_InitStructure;
+  NVIC_InitTypeDef   NVIC_InitStructure;
+
+  gpio_clock(gpio_pin);
+
+  /* Configure gpio_pin as input floating */
+  GPIO_InitStructure.GPIO_Pin = gpio_pin & GPIO_PIN_MASK;
```

Comment:
Alright so anytime we use a gpio pin and want to know that actual bits you have to chop off the first 16 bits which are used to identify which GPIO. The bottom 16 are the same as `GPIO_Pin_X`, therefore we use a `GPIO_PIN_MASK` which is the lower 16 bits and bitwise and them together whenever we want the `GPIO_Pin_X` value.